File: backend/app/core/hexagram.py
import random
from typing import List, Optional, TypedDict

class HexagramReading(TypedDict):
    name: str
    chinese: str
    description: str
    judgment: str
    image: str
    lines: dict[str, str]

class Hexagram:
    def __init__(self) -> None:
        self.lines: List[int] = []
        self.changing_lines: List[int] = []
        self.number: Optional[int] = None

    def _divide_stalks(self) -> int:
        total = 0
        for _ in range(3):
            total += random.choice([2, 3])
        return total

    def generate(self) -> None:
        self.lines = []
        self.changing_lines = []

        for _ in range(6):
            total = self._divide_stalks()
            if total == 6:
                self.lines.append(0)
                self.changing_lines.append(len(self.lines))
            elif total == 9:
                self.lines.append(1)
                self.changing_lines.append(len(self.lines))
            elif total == 7:
                self.lines.append(1)
            elif total == 8:
                self.lines.append(0)

        self.number = self._calculate_hexagram_number()

    def _calculate_hexagram_number(self) -> int:
        binary = "".join(str(line) for line in reversed(self.lines))
        decimal = int(binary, 2)
        return decimal + 1

    def get_relating_hexagram(self) -> Optional["Hexagram"]:
        if not self.changing_lines:
            return None

        relating_lines = list(self.lines)

        for changing_line_index in self.changing_lines:
            relating_lines[changing_line_index - 1] = 1 - relating_lines[changing_line_index - 1]

        relating_hexagram = Hexagram()
        relating_hexagram.lines = relating_lines
        relating_hexagram.number = relating_hexagram._calculate_hexagram_number()

        return relating_hexagram 

File: packages/frontend/src/types/hexagram.ts
export interface Trigram {
  name: string;
  meaning: string;
  image: string;
  family: string;
  attribute: string;
}

export interface LineInfo {
  lineNumber: number;
  type: "yang" | "yin";
  meaning: string;
}

export interface TrigramSignificance {
  upper: string;
  lower: string;
  relationship: string;
}

export interface HexagramData {
  number: number;
  name: string;
  chineseName: string;
  pinyin?: string;
  englishName?: string;
  chinese?: string;
  description?: string;
  judgment: string | { text?: string };
  image: string | { text?: string };
  lines?: LineInfo[];
  upperTrigram?: string;
  lowerTrigram?: string;
  trigramSignificance?: TrigramSignificance;
  commentary?: string[];
}

export interface ReadingResponse {
  hexagram_number: number;
  changing_lines: number[];
  lines: string[];
  reading: HexagramData;
  relating_hexagram?: HexagramData | null;
}

export interface Hexagram {
  number: number;
  name: string;
  chinese: string;
  description: string;
  judgment: {
    text: string;
    explanation?: string;
  };
  image: {
    text: string;
    explanation?: string;
  };
  lines: string[];
  trigrams: {
    upper: string;
    lower: string;
  };
}

export interface HexagramLine {
  position: number;
  value: number;
  isChanging: boolean;
  text?: string;
}

export type HexagramMode = "yarrow" | "coin" | "random";

// Types for I Ching hexagrams and readings

export interface Trigrams {
  upper: string;
  lower: string;
}

export interface HexagramData {
  number: number;
  name: string;
  description: string;
  judgment?: string;
  image?: string;
  trigrams?: Trigrams;
}

export interface HexagramReading {
  hexagram_number: number;
  lines: string[];
  changing_lines: number[];
  reading: HexagramData;
  relating_hexagram?: HexagramData | null;
  interpretation?: string;
}

export enum HexagramLine {
  YIN = 0, // Broken line
  YANG = 1, // Solid line
}

export interface ReadingRequest {
  question: string;
  mode?: string;
  verbose?: boolean;
  seed?: number;
}

export enum DivinationMethod {
  YARROW_STALKS = "yarrow_stalks",
  THREE_COINS = "coins",
  MANUAL = "manual",
}
